### PR TITLE
[terra-table] Add row selection to table component

### PR DESCRIPTION
**Pull request by cm9361**
### Summary

**What was changed:**
The Terra Table was updated to allow consumers to enable the row selection mode and its corresponding capabilities.

**Why it was changed:**
Although an accessible table component would not normally have row selection, the capability is required in the Terra Table component for passivity and to allow integration in our clinical workflows.

### Testing

This change was tested using:

- [X] WDIO
- [X] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [X] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details
N/A

**This PR resolves:**

UXPLATFORM-9766 

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

**Review by adoroshk — APPROVED** [review]
Tested locally (attaching a screenshot). Looks good to me. As the jest failure looks like not related to current code change and is coming from the terra-menu component, approving the change.
Screenshot 2023-10-25 at 1 22 21 PM

**Comment by adoroshk** [comment]
It looks like the failing jest test issue is being addressed in this PR: https://github.com/cerner/terra-framework/pull/1855/files#diff-038f1481da2f8f54dfaea6e6dd47bd1e2dfc0d58823c972fde3900f1663dc71f

File: packages/terra-table/src/Table.jsx [hunk 1]
```diff
@@ -342,14 +413,16 @@ function Table(props) {
                 hasRowSelection={hasSelectableRows}
                 displayedColumns={displayedColumns}
                 rowHeaderIndex={rowHeaderIndex}
-                onCellSelect={isGridContext ? handleCellSelection : undefined}
+                onCellSelect={isGridContext || hasSelectableRows ? handleCellSelection : undefined}
```
Comment (smason0) [inline packages/terra-table/src/Table.jsx:416]: I might be missing why this condition was added here. Was this added as part of allowing the row selection cell checkbox to be selectable?

File: packages/terra-framework-docs/src/terra-dev-site/doc/table/Examples/TableRowSelection.jsx [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        { content: 'Inpatient, 9 months' },
+        { content: '' },
+        { content: 'Quinzell, Harleen' },
+        { content: '' },
+      ],
+    },
+  ],
+};
+
+const RowSelection = () => {
```
Comment (smason0) [inline packages/terra-framework-docs/src/terra-dev-site/doc/table/Examples/TableRowSelection.jsx:176]: I would have expected a test page similar to this one and some accompanying WDIO tests for row selection.

File: packages/terra-table/src/Table.jsx (hunk 1 above)
Comment (cm9361) [reply packages/terra-table/src/Table.jsx:416]: Previously the onCellSelect, which lets us know that a cell was "selected", would only trigger when in the grid context.  If we are in row selection mode, hasSelectableRows prop equal to true, then we need to also handle the cell selection.

File: packages/terra-framework-docs/src/terra-dev-site/doc/table/Examples/TableRowSelection.jsx (hunk 2 above)
Comment (cm9361) [reply packages/terra-framework-docs/src/terra-dev-site/doc/table/Examples/TableRowSelection.jsx:176]: I agree with you.  I covered in jest, but should also have added the appropriate WDIO tests.  Thank you for catching this.

**Comment by chrismichalewicz** [comment]
@cm9361 I did a functional review as well as tested keyboard and screenreader usage. I thought it looked good, great job!

Mouse: Able to select check boxes as expected. Noticed the row color change on mouse hover, nice touch as that will address one of the MPages gaps identified!

Keyboard: Able to navigate the checkboxes via tab and select/unselect them via Space. Rows selected/unselected as expected. 

Screenreader: Tested with Edge and JAWS 2022.2202.38
Row selection mode check box was communicated accurately along with instructions to use that worked. 
When individual row check boxes had focus Row header was communicated along with check box. 
Selecting the check box via space bar effectively communicated that the row was selecting or unselected.

**Review by smason0 — APPROVED** [review]

File: packages/terra-framework-docs/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,7 @@
 ## 1.43.0 - (October 25, 2023)
 
 * Added
+  * Added examples and test to `terra-table` for row selection mode.
```
Comment (sdadn) [inline packages/terra-framework-docs/CHANGELOG.md:8]: Can you move this to the released section?

Also looks like these entries for `packages/terra-framework-docs/CHANGELOG.md` also got merged in the wrong section:

https://github.com/cerner/terra-framework/pull/1858

**Review by eawww — APPROVED** [review]
Noticed in this review that the elements being focused in the selectable table headers, for Table specifically, are the header cells rather than the buttons inside of them. Logged [this jira](https://jira2.cerner.com/browse/UXPLATFORM-9801) to remedy.

This one looks good for the story though.

**Pull request merged**